azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -134,15 +136,45 @@ public void cleanUpStaleFlows() {
   }
 
   /**
-   * Try to clean the stale containers that are older than maximum azkaban flow running time
+   * From all pods in the current namespace and Azkaban cluster, select the pods whose executions
+   * are not in unfinished statuses, e.g. DISPATCHING, PREPARING, RUNNING, PAUSED, KILLING,
+   * FAILED_FINISHING in DB, and delete the pods & services to release resources
    */
-  public void cleanUpStaleContainers() {
-    Duration containerValidity = validityMap.get(Status.RUNNING).getFirst();
-    if (!containerValidity.isNegative()) {
+  public void cleanUpContainersInTerminalStatuses() {
+    logger.info("Cleaning up pods of terminated flow executions");
+    Set<Integer> activeFlows = new HashSet<>();
+    Set<Integer> currentNameSpacedPods = new HashSet<>();
+    ImmutableMap<Status, Pair<Duration, String>> unFinishedStatusesMap =
+        new Builder<Status,
+        Pair<Duration, String>>()
+        .put(Status.DISPATCHING, new Pair<>(Duration.ZERO, SUBMIT_TIME))
+        .put(Status.PREPARING, new Pair<>(Duration.ZERO, SUBMIT_TIME))
+        .put(Status.RUNNING, new Pair<>(Duration.ZERO, START_TIME))
+        .put(Status.PAUSED, new Pair<>(Duration.ZERO, START_TIME))
+        .put(Status.KILLING, new Pair<>(Duration.ZERO, UPDATE_TIME))
+        .put(Status.FAILED_FINISHING, new Pair<>(Duration.ZERO, START_TIME))
+        .build();
+    // Obtain all pods in current namespace which are dispatched from current Azkaban cluster
+    try {
+      currentNameSpacedPods = this.containerizedImpl.getContainersByDuration(Duration.ZERO);
+    } catch (final ExecutorManagerException e) {
+      logger.error("Unable to obtain the pods in the current namespace and Azkaban cluster", e);
+    }
+    // Obtain all containerized executions that are in unfinished statuses
+    for (Status status : unFinishedStatusesMap.keySet()) {
       try {
-        this.containerizedImpl.deleteAgedContainers(containerValidity);
-      } catch (final Exception e) {
-        logger.error("Exception occurred while cleaning up stale pods and services." + e);
+        List<ExecutableFlow> flows = (this.executorLoader.fetchStaleFlowsForStatus(status,
```

Comment:
```suggestion
        List<ExecutableFlow> flows = this.executorLoader.fetchStaleFlowsForStatus(status,
```
And remove the ")" in the next line

Reply:
Removed redundant brackets.